docs/RFC/draft-paragonie-paseto-rfc-00.txt

```diff
@@ -0,0 +1,1680 @@
+   [RFC8017], with 2048-bit private keys.  Refer to the operations
+   defined in *v1.Sign* and *v1.Verify* for a formal definition.
+
+4.3.  Version v1 Algorithms
+
+4.3.1.  v1.GetNonce
+
+   Given a message ("m") and a nonce ("n"):
+
+   1.  Calculate HMAC-SHA384 of the message "m" with "n" as the key.
+
+   2.  Return the leftmost 32 bytes of step 1.
+
+4.3.2.  v1.Encrypt
+
+   Given a message "m", key "k", and optional footer "f" (which defaults
+   to empty string):
+
+   1.  Set header "h" to "v1.local."
+
+   2.  Generate 32 random bytes from the OS's CSPRNG.
+
+   3.  Calculate "GetNonce()" of "m" and the output of step 2 to get the
+       nonce, "n".
+
+       *  This step is to ensure that an RNG failure does not result in
+          a nonce-misuse condition that breaks the security of our
+          stream cipher.
+
+   4.  Split the key ("k") into an Encryption key ("Ek") and an
+       Authentication key ("Ak"), using the leftmost 16 bytes of "n" as
+       the HKDF salt.  (See below for pseudocode.)
+
+   5.  Encrypt the message using "AES-256-CTR", using "Ek" as the key
+       and the rightmost 16 bytes of "n" as the nonce.  We'll call this
+       "c".  (See below for pseudocode.)
+
+   6.  Pack "h", "n", "c", and "f" together (in that order) using PAE
+       (see Section 2.2).  We'll call this "preAuth".
+
+   7.  Calculate HMAC-SHA384 of the output of "preAuth", using "Ak" as
+       the authentication key.  We'll call this "t".
+
+   8.  If "f" is:
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018                [Page 7]
+
+Internet-Draft                                                April 2018
+
+
+       *  Empty: return h || b64(n || c || t)
+
+       *  Non-empty: return h || b64(n || c || t) || "." || b64(f)
+
+       *  ...where || means "concatenate"
+
+   Example code:
+
+                   Ek = hkdf_sha384(
+                      len = 32
+                      ikm = k,
+                      info = "paseto-encryption-key",
+                      salt = n[0:16]
+                   );
+                   Ak = hkdf_sha384(
+                      len = 32
+                      ikm = k,
+                      info = "paseto-auth-key-for-aead",
+                      salt = n[0:16]
+                   );
+
+              Step 4: Key splitting with HKDF-SHA384 as per .
+
+                          c = aes256ctr_encrypt(
+                              plaintext = m,
+                              nonce = n[16:]
+                              key = Ek
+                          );
+
+               Step 5: PASETO v1 encryption (calculating c)
+
+4.3.3.  v1.Decrypt
+
+   Given a message "m", key "k", and optional footer "f" (which defaults
+   to empty string):
+
+   1.  If "f" is not empty, verify that the value appended to the token
+       matches "f", using a constant-time string compare function.  If
+       it does not, throw an exception.
```

Comment:
> Thinking about this footer validation, I'm realizing I'm confused why it's necessary. Are there not two options for where "f" comes from:
>
> 1. it's pulled off the token, in which case it'll always match, or;
> 2. it's known to the recipient, in which case it's unnecessary to send?

This was something I was meaning to ask about. My implementation exclusively does `1.` with verification happening cryptographically ([e.g.](https://github.com/aidantwoods/swift-paseto/blob/5fdffc428e3d69b47de5ce3d769ba640fa773f2b/Sources/Paseto/Implementations/Version2.swift#L100-L121)). I don't currently allow users to specify an expected footer – it's either part of the received token or it isn't. I'm curious to know if I should be adding this.

Reply:
The entire point of the footer was to allow `key-id` claims to be stored without encryption.

Allowing users to pull the footer before verifying the claim isn't dangerous, as long as they do eventually pass this footer value back to the `decode` step.

If we weren't supporting `key-id` use-cases (which I personally hate but a lot of people use it for key rotation schemes so it's a good idea to meet their needs), this isn't a big deal: The footer will never be populated, and the spec can be simplified.

So with that in mind, yes, the ability to fetch the footer before authenticating/decrypting the token should be a feature. Furthermore, I think the logic here might be a bit paranoid about people abusing the footer, so we should be able to relax these constraints.

Reply:
> So with that in mind, yes, the ability to fetch the footer before authenticating/decrypting the token should be a feature.

Supposing that the user does pull the footer out, I guess the question is why do we verify the footer that the user has given us, as opposed to just verifying the one attached to the message? Is this to try to make sure the user does not mix up footers?

i.e. why do we do
```
f = extractFooter(m)
v1.decrypt(m, k, f)
```
where `m` contains the footer, but we check it matches. Instead of

```
f = extractFooter(m)
v1.decrypt(m, k)
```

Where `v1.decrypt` does not accept a footer argument, but determines `f` itself and performs only a cryptographic check to ensure the integrity of `m`. The only difference I can see here is that the user might not give us the right footer back.